ROS User Guide -  XRCE-DDS update (#2259)
PX4/PX4-user_guide: en/middleware/xrce_dds.md

# XRCE-DDS (PX4-FastDDS Bridge)

:::note
XRCE-DDS replaces the [Fast-RTPS Bridge](https://docs.px4.io/v1.13/en/middleware/micrortps.html#rtps-dds-interface-px4-fast-rtps-dds-bridge) used in PX4 v1.13.
:::

> PX4 uses XRCE-DDS middlware to allow [uORB messages](../middleware/uorb.md) to be published and subscribed on a companion computer as though they were ROS 2 topics.

Author: ```suggestion
PX4 uses XRCE-DDS middleware to allow [uORB messages](../middleware/uorb.md) to be published and subscribed on a companion computer as though they were ROS 2 topics.
```

This provides a fast and reliable integration between PX4 and ROS2, and makes it much easier for ROS 2 applications to get vehicle information and send commands. 

PX4 uses an XRCE-DDS implementation that leverages [eProsima Micro XRCE-DDS](https://micro-xrce-dds.docs.eprosima.com/en/stable/introduction.html).

The following guide describes the architecture and various options for setting up the client and agent.
In particular it covers the options that are most important to PX4 users.

> 

Author: FYI Point  above is that this is not an exhaustive guide - for that you see the eProsima guide. But we don't do everything the architecture allows - e.g. connection over CAN, so this is "PX4s take on the infrastructure".

## Architecture

> The XRCE-DDS middleware consists of a client running on PX4 and an agent running on the companion computer, with bi-directional data exchange between them over a serial, UDP, TCP or custom link.

Author: Do we support these options:  serial, UDP, TCP ? What about CAN and the other links that eProsima say they support?

Reviewer: @hamishwillee PX4 supports serial and UDP on the client side at the moment.

Author: ```suggestion
The XRCE-DDS middleware consists of a client running on PX4 and an agent running on the companion computer, with bi-directional data exchange between them over a serial or UDP link.
```

On Ubuntu you can build from source and install the Agent standalone using the following commands:

```sh
git clone https://github.com/eProsima/Micro-XRCE-DDS-Agent.git
cd Micro-XRCE-DDS-Agent
mkdir build
cd build
cmake ..
make
sudo make install
sudo ldconfig /usr/local/lib/
```

:::note
There are various build configuration options linked from the corresponding topic in the [official guide]( https://micro-xrce-dds.docs.eprosima.com/en/latest/installation.html#installing-the-agent-standalone), but these have not been tested.
:::

To start the agent with settings for connecting to the XRCE-DDS client running on the simulator:

```sh
MicroXRCEAgent udp4 -p 8888
```

### Install from Snap Package

Install from a snap package on Ubuntu using the following command:

```sh
sudo snap install micro-xrce-dds-agent --edge
```

To start the agent with settings for connecting to the XRCE-DDS client running on the simulator (note that the command name is different than if you build the agent locally):

```sh
micro-xrce-dds-agent udp4 -p 8888
```

:::note
At time of writing the stable of version installed from snap connects to PX4 but reports errors creating topics.

> The development version, fetched using `--edge` above, does work.

Author: FYI - `stable` doesn't work for me.

:::

### Build/Run within ROS2 Workspace

The agent can be built and launched within a ROS2 workspace (or build standalone and launched from a workspace.

> You must already have installed ROS 2 following the instructions in: [ROS 2 User Guide > Install ROS 2](##TBDXXX)

Author: - [ ] Need to fix up this link

Author: ```suggestion
You must already have installed ROS 2 following the instructions in: [ROS 2 User Guide > Install ROS 2](../ros/ros2_comm.md#install-ros-2).
```

To run the XRCE-DDS agent in the workspace:

1. Source the `local_setup.bash` to make the executables available in the terminal (also `setup.bash` if using a new terminal).

   ```sh
   source /opt/ros/foxy/setup.bash
   source install/local_setup.bash
   ```
1. Start the agent with settings for connecting to the XRCE-DDS client running on the simulator: 

   ```sh
   MicroXRCEAgent udp4 -p 8888
   ```

## Starting XRCE-DDS

### Starting the Agent

The agent is used to connect to the client over a particular channel, such as UDP, TCP, or a serial connection.
The channel settings are specified when the agent is started, using command line options. 
These are documented in the eProsima user guide: [Micro XRCE-DDS Agent > Agent CLI](https://micro-xrce-dds.docs.eprosima.com/en/latest/agent.html#agent-cli). <!-- what options work? i.e. does PX4 allow CAN? -->

:::note
You should create a single instance of the agent for each channel over which you need to connect.
:::

For example, the PX4 simulator runs the XRCE-DDS client over UDP on port 8888, so to connect to the simulator you would start the agent with the command:

```sh

> MicroXRCEAgent udp4 -p 8888

Author: This is all a bit "thin". Can't really update until we have a working companion setup guide/example.

```

> ### Starting the Client

Reviewer: There is one last method to start the client on real vehicles: using the parameter `XRCE_DDS_0_CFG` of the group [Micro XRCE-DDS](https://docs.px4.io/main/en/advanced_config/parameter_reference.html#micro-xrce-dds). With that you can start the client on serial or Ethernet or even WiFi (it seems). Then you have other parameters, like `XRCE_DDS_UDP_PRT` to setup the UDP port, `XRCE_DDS_DOM_ID` if you want to connect to a different `DOMAIN_ID` (https://docs.ros.org/en/foxy/Concepts/About-Domain-ID.html) and finally `XRCE_DDS_KEY` if you want to change the default `SESSION_ID` (needed for multi vehicle simulations). Anyway, I think the last two parameters can be put in the docs later on.

Author: @beniaminopozzan Thanks! Actually you should ALWAYS use the parameters on hardware. If you set the command you're setting yourself up for conflicts with other configuration. I have mentioned it still, but not as recommended.

I have rewritten this section, can you check?

Reviewer: Thanks!!!

For serial connection I think it is perfect, for Ethernet connection it seems that @ozitoune managed to have it working on Pixhawk 6X with
```
microdds_client start -t udp -p 8888 -h 192.168.0.100
```
where `-h 192.168.0.100` tells the IP address of the _Agent_. As this is not a parameter in PX4, it is then required to start the client manually or from System Startup.

Author: @beniaminopozzan That's horrible - can you create an issue to get a parameter for the host into PX4?

While most releases should support a very similar set of messages, to be certain you would need to check the yaml file for your particular release.
<!-- Jublish the set we use?: https://github.com/PX4/px4_msgs/issues/22 -->

Note that ROS 2/DDS needs to have the _same_ message definitions that were used to create the XRCE-DDS client module in the PX4 Firmware in order to interpret the messages:

- If you're using a main or release version of PX4 you can get these by cloning [PX4/px4_msgs](https://github.com/PX4/px4_msgs) into your workspace (branches are provided for each release).
- If you're creating your own new topic you will need to update [dds_topics.yaml](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/microdds_client/dds_topics.yaml) and also copy the message definitions out of PX4 source tree into any ROS2 workspace that needs them.
  The messages must be in a folder named `px4_msgs/msg/`.
  
  :::note
  Technically the messages must be in a folder defined by the `type` prefix in the yaml file:
  As you can see below, this is `px4_msgs::msg::`:
  
  ```yaml
  - topic: /fmu/out/vehicle_odometry
    type: px4_msgs::msg::VehicleOdometry
  ```
  :::

> ### PX4 ROS2 QoS Settings

Reviewer: No, I agree that it is convenient to a quick reference on the QoS details used by the client!

PX4 QoS settings for publishers are incompatible with the default QoS settings for ROS2 subscribers.
So if ROS2 code needs to subscribe to a uORB topic, it will need to use compatible QoS settings.

> One example of which is shown in [ROS 2 User Guide > ROS2 Subscriber QoS Settings]()

Author: - [ ] Fix link

Author: ```suggestion
One example of which is shown in [ROS 2 User Guide > ROS2 Subscriber QoS Settings](../ros/ros2_comm.md#ros2-subscriber-qos-settings).
```

  .history = UXR_HISTORY_KEEP_LAST,
  .depth = 0,
};
```

PX4 uses the following QoS settings for subscribers:

```cpp
uxrQoS_t qos = {
  .durability = UXR_DURABILITY_VOLATILE,
  .reliability = UXR_RELIABILITY_BEST_EFFORT,
  .history = UXR_HISTORY_KEEP_LAST,
  .depth = queue_depth,
};
```

ROS2 uses the following QoS settings (by default) for publishers and subscriptions: "keep last" for history with a queue size of 10, "reliable" for reliability, "volatile" for durability, and "system default" for liveliness.
Deadline, lifespan, and lease durations are also all set to "default".
<!-- From https://github.com/PX4/PX4-user_guide/pull/2259#discussion_r1099788316 -->

## Helpful Resources

- [ROS2 in PX4: ROS2 in PX4: Technical Details of a Seamless Transition to XRCE-DDS](https://www.youtube.com/watch?v=F5oelooT67E) - Pablo Garrido & Nuno Marques (youtube)
- [PX4 ROS2 offboard tutorial](https://gist.github.com/julianoes/adbf76408663829cd9aed8d14c88fa29) (Github gist - JulianOes)
- [ROS2 PX4 Offboard Tutorial](https://github.com/Jaeyoung-Lim/px4-offboard/blob/2d784532fd323505ac8a6e53bb70145600d367c4/doc/ROS2_PX4_Offboard_Tutorial.md) (Jaeyoung-Lim).

<!---

> Some of this might be useful.

Author: Below here is stuff from the old architecture on uart config. I've left here - but probably would like in a "hardware tutorial".

   ```sh
   source /opt/ros/foxy/setup.bash
   source install/local_setup.bash
   ```
1. Start the agent with settings for connecting to the XRCE-DDS client running on the simulator: 

   ```sh
   MicroXRCEAgent udp4 -p 8888
   ```

## Starting XRCE-DDS

### Starting the Agent

The agent is used to connect to the client over a particular channel, such as UDP, TCP, or a serial connection.
The channel settings are specified when the agent is started, using command line options. 
These are documented in the eProsima user guide: [Micro XRCE-DDS Agent > Agent CLI](https://micro-xrce-dds.docs.eprosima.com/en/latest/agent.html#agent-cli). <!-- what options work? i.e. does PX4 allow CAN? -->

:::note
You should create a single instance of the agent for each channel over which you need to connect.
:::

For example, the PX4 simulator runs the XRCE-DDS client over UDP on port 8888, so to connect to the simulator you would start the agent with the command:

```sh
MicroXRCEAgent udp4 -p 8888
```

On RaspberryPi, if you're using a serial link and connected to the IO pins, you might connect using this command:

```sh
sudo MicroXRCEAgent serial --dev /dev/AMA0 -b 921600

> ```

Author: @beniaminopozzan I added this - seem correct to you? Based on discussion in https://github.com/PX4/PX4-Autopilot/issues/21120#issuecomment-1428347091

Reviewer: I didn't have time to go through all those  "issue" topics on the PX4 repository yet. In particular I'm not 100% sure that /dev/AMA0 are the TX RX pins of the raspberry. And if you are using Raspberry Pi 4, which has multiple serial ports things gets more complicated. I'll get back to you as soon as possible

Reviewer 2: Hi, I have a step-by-step tutorial that describe this raspberry setup. Maybe it can help somehow:

https://github.com/ViniciusAbrao/px4_ros2_xrcedds/blob/master/raspiberrypi_setup.pdf

Author: So the short answer is "Yes" we should link to the companion computer setup, and either expand it, or possibly have a RaPI setup page below it. This needs to cross link to the official RaPi docs.

I will think about minimal amount to put in here on Wednesday and merge this. Vince, I am sure your doc will be useful - I'll check it out Weds.

@beniaminopozzan Thanks so much for your help!

- [XRCE_DDS_0_CFG](../advanced_config/parameter_reference.md#XRCE_DDS_0_CFG): Set the port to connect on, such as `TELEM2`, `Ethernet`, or `Wifi`.
  - [XRCE_DDS_UDP_PRT](../advanced_config/parameter_reference.md#XRCE_DDS_UDP_PRT):
    If using an Ethernet connect, use this to specify the UDP port.
  - [SER_TEL2_BAUD](../advanced_config/parameter_reference.md#SER_TEL2_BAUD), [SER_URT6_BAUD](../advanced_config/parameter_reference.md#SER_URT6_BAUD) (and so on):
    If using a serial connection, the baud rate must be set using the `_BAUD` parameter associated with the serial port.
    For example, you'd set a value for `SER_TEL2_BAUD` if you are connecting to the companion using `TELEM2`.
    For more information see [Serial port configuration](../peripherals/serial_configuration.md#serial-port-configuration).

  Some setups might also need these parameters to be set:

  - [XRCE_DDS_KEY](../advanced_config/parameter_reference.md#XRCE_DDS_KEY): The XRCE-DDS key.
    If you're working in a multi-client, single agent configuration, each client should have a unique key.

>   - [XRCE_DDS_DOM_ID](../advanced_config/parameter_reference.md#XRCE_DDS_DOM_ID): TBD.

Author: TBD could be added as a post process - but what is it for/how/when should it be set?

Reviewer: - `XRCE_DDS_KEY` sets the `key` value of the client _session_. All clients connected to the same agent must have different keys (see https://micro-xrce-dds.docs.eprosima.com/en/stable/client_api.html#session , `uxr_init_session`). Moreover the key cannot have value zero. It is important to set this parameter only for multi vehicle setups and, more importantly, multi vehicle simulations, where all clients are connected in UDP to the same agent (multipla agents operating on different ports could be used but I don't se any advantages in doing that)

- `XRCE_DDS_DOM_ID` is the equivalent of the [`ROS_DOMAIN_ID`](https://docs.ros.org/en/foxy/Concepts/About-Domain-ID.html) and gives logical separation between DDS networks. Therefore clients configured with different `XRCE_DDS_DOM_ID` will belong to different networks. By default, ROS2 operates on ID 0

> # XRCE-DDS: PX4-FastDDS Bridge

Author: At moment just a scratchpad of topics I'll need.

MicroXRCEAgent udp4 -p 8888
```

When working with real hardware, the setup depends on the hardware, OS, and channel.
For example, if you're using the RasPi `UART0` serial port, you might connect using this command (based on the information in [Raspberry Pi Documentation > Configuring UARTS](https://www.raspberrypi.com/documentation/computers/configuration.html#configuring-uarts)):

```sh
sudo MicroXRCEAgent serial --dev /dev/AMA0 -b 921600
```

:::note
For more information about setting up communications channels see [Pixhawk + Companion Setup > Serial Port setup](../companion_computer/pixhawk_companion.md#serial-port-setup), and sub-documents.

> :::

Author: @beniaminopozzan @ViniciusAbrao This generalizes the info about agent startup, making it clear that "it depends", linking to the RaPi configuring UARTs to help people understand a bit more about this case.
I also link, as suggested, to the companion computers page. That has some general Linux setup.

@ViniciusAbrao What would be great is if you could add your setup information as a topic below https://docs.px4.io/main/en/companion_computer/pixhawk_companion.html 

If this can be general it would be "Raspberry Pi Companion". If it has to be specific to a particular RaPi then ""Raspberry Pi Xxxx Companion", where Xxxx is the version. I think it can be general, but we'd need notes to highlight things to check in a particular system.

What would be great is more than just this specific part of setup around ROS. RasPi is very popular, so would be good to have images of the hardware connection showing the lines. Explanation of how it is powered, and why done that way, etc.
The setup of the connection could then be part of this.

Does it make sense/is it possible?

Reviewer: It definitively makes sense. The linked docs covers only for USB-to-serial adapter but most (not only Raspberry Pi) companion computers have hardware serial devices.

Author: Yes, it's a simplistic topic. In an ideal world you'd start with a doc that covers all connection scenarios for as much hardware and OS as possible. That requires someone with broad experience commit quite a lot of time. Getting support for that is hard.

So the way I approach these problems is to attempt to build out specific hardware scenarios, then generalize them.
So for example here if @ViniciusAbrao can create a RaPi Companion doc covering as much as possible about the connections then that might contain information about non-USB serial connections, which can then migrate to the parent doc. 

Thats my thinking, anyway.

From that we do things like potentially cross link to https://docs.px4.io/main/en/companion_computer/holybro_pixhawk_rpi_cm4_baseboard.html with more software setup instructions. The idea being that we make setup of any RaPi setup as seamless as possible. 

@ViniciusAbrao You onboard to create that doc? RaPi is a very popular option, so I think a "good start". One thing no one seems to cover on companion setups, at least in detail, is how to power the RaPi.

Reviewer 2: Yes @hamishwillee ... I am going to create the doc covering the connections and the setup. I let you know when I finish it !!

Author: Thanks @ViniciusAbrao - awesome!!!! Any kind of ETA?

Reviewer 2: > Thanks @ViniciusAbrao - awesome!!!! Any kind of ETA?

I am going to work on it until next week !!

Reviewer 2: @hamishwillee I have finished the review. Please, find the PR at:

https://github.com/PX4/PX4-user_guide/pull/2318#issuecomment-1451006165

If it is not the best way to push it, please let me know. Also, maybe the English can be improved, so, feel free to edit with corrections or technical changes.

Author: Thanks @ViniciusAbrao - I've had a first look at it and asked some questions. Next week I'll look at your answers, restructure it to be more easy to extend later (for example if we add information about "how to connect via Ethernet") and link it in.

Would be good if you could push it as a branch other than main in the PR sometime in the next week - that makes editing much easier with my knowledge of git.

Reviewer 2: Thanks @hamishwillee The new PR is done at:

https://github.com/PX4/PX4-user_guide/pull/2323

I answered your questions... please, let me know if some points still not clear.

  See [Serial port configuration](../peripherals/serial_configuration.md#serial-port-configuration).
:::

Once set, you may need to reboot PX4 for the parameters to take effect.
They will then perist through subsequent reboots.

While not recommended, you can also start the [microdds-client](../modules/modules_system.md#microdds-client) using a command line.
This can be called as part of [System Startup](../concept/system_startup.md) or through the [MAVLink Shell](../debug/mavlink_shell.md) (or a system console).
For example, the following command can be used to connect via Ethernet to a remote host at `192.168.0.100:8888`.

```sh
microdds_client start -t udp -p 8888 -h 192.168.0.100
```

:::note
At time of writing there is no PX4 parameter for setting the address of the remote host on which the XRCE-DDS agent is running.
Therefore for Ethernet connections you will _have_ to use a command to start the client.

> :::

Author: FYI @beniaminopozzan This "fixes" the docs to explain you need to use command line for Ethernet, as discussed in https://github.com/PX4/PX4-user_guide/pull/2259/commits/986995c6a4309cc496d17c67659a336ded449045#r1108235444

If you give me an issue for creating the host parameter for pointing to the agent I'll add that as a link from "at the time of writing" in the note.

Reviewer: Here it is!
https://github.com/PX4/PX4-Autopilot/issues/21179

Author: Thanks!!! I'll add to docs.
